src/main/java/com/example/gradecalculator/service/SubjectService.java

```diff
@@ -0,0 +1,153 @@
+            long selectedYearId = Long.parseLong(splittedString[1]);
+            String error = saveSubjectYearSelectedUser(userId ,selectedSubjectId, selectedYearId);
+            if(error != null){
+                errors.add(error);
+            }
+        }
+
+        return errors;
+    }
+
+    public String saveSubjectYearSelectedUser(long userId, long subjectId, long yearId) {
+
+        Subject selectedSubject = subjectRepository.findById(subjectId).orElseThrow(() -> new IllegalArgumentException("Subject not found"));
+        SchoolYear selectedYear = schoolYearRepository.findById(yearId).orElseThrow(() -> new IllegalArgumentException("Year not found"));
+        User selectedUser = userRepository.findById(userId).orElseThrow(() -> new IllegalArgumentException("User not found"));
+        this.selectSubjectForYear(selectedYear.getStartDate().getYear(), selectedSubject.getName());
+        var subjects = userSubjectRepository.findByUserId(userId);
+        boolean exists = subjects.stream().anyMatch(item -> item.getSubject().getId().equals(subjectId) && item.getSchoolYear().getId().equals(yearId));
+        if (!exists) {
+            UserSubject userSubject = new UserSubject();
+            userSubject.setUser(selectedUser);
+            userSubject.setSubject(selectedSubject);
+            userSubject.setSchoolYear(selectedYear);
+            userSubjectRepository.save(userSubject);
+            return null;
+        }
+        return ("Subject already selected for the given year: " + selectedSubject.getName() + " - " + selectedYear.getName());
+    }
+
+    public TreeMap<String, Set<UserSubjectTO>> selectedSubject(long userId) {
+
+        var userSubjects = userSubjectRepository.findByUserId(userId);
+        TreeMap<String, Set<UserSubjectTO>> subjectsByYear = new TreeMap<>();
+        for (UserSubject userSubject : userSubjects) {
+            String year = userSubject.getSchoolYear().getName();
+            var userSubjectTO = new UserSubjectTO();
+            userSubjectTO.setID(userSubject.getId());
+            userSubjectTO.setName(userSubject.getSubject().getName());
+            subjectsByYear.computeIfAbsent(year, k -> new HashSet<>()).add(userSubjectTO);
+        }
+
+        return subjectsByYear;
+    }
+
+
+    @Transactional
+    public boolean deleteSubject(Long subjectId, String userId) {
+        Optional<UserSubject> userSubjectOpt = userSubjectRepository.findById(subjectId);
+
+        if (userSubjectOpt.isPresent()) {
+            UserSubject userSubject = userSubjectOpt.get();
+            if (userSubject.getUser().getId().equals(Long.parseLong(userId))) {
+                userSubjectRepository.delete(userSubject);
+                boolean exists = userSubjectRepository.existsById(subjectId);
+                if (!exists) {
+                    System.out.println("Subject with ID " + subjectId + " successfully deleted.");
+                    return true;
+                } else {
```

Comment:
there is no need to double check if the entry is deleted. the delete action will already throw an error if it fails